Revamp the `Web Applications` page (#2692)
apache/logging-log4j2: src/site/antora/modules/ROOT/pages/manual/webapp.adoc

. Each application can use their own copy of Log4j Core and include `log4j-core` in the WAR or EAR archive.
. Applications can also use a single copy of Log4j Core that must be installed globally on the application server.

While the first approach is the easiest to implement, it has some limitations:

* Log events emitted by each application and the libraries bundled with it will be handled by Log4j Core,
but events related to the application emitted by a **shared** library (e.g. JPA implementation) will be handled by the application server.
To diagnose a problem with the application you might need to look into multiple log files.
* Each application must use a **different** log file to prevent problems with concurrent access to the same file by multiple applications.
Problems may arise especially if xref:manual/appenders.adoc#rollingfileappender[rolling file appender] is used.
* Web applications have a different lifecycle from the application server.
Additional care is required to stop Log4j Core when the application is stopped.
See <<log4j-jakarta-web>> for more details.

The second approach is more complex and requires changes to the configuration of the application server.

> See <<sharing>> for more details.

Reviewer: The way 2nd approach is introduced strongly implies a preference towards the 1st one. On the contrary, I personally find the 1st approach a bad practice. Judging from my experience, logging should be ideally configured centrally in an application server. As a matter of fact, that is what I have seen most in companies. Am I mistaken with this observation?

[#log4j-jakarta-web]

> == Integrating with Web Applications

Reviewer: ```suggestion
== Integrating with web applications
```

[NOTE]
====
To avoid problems, some Log4j API and Log4j Core features are automatically disabled, when running in a Jakarta EE environment. Most notably:

- the usage of `ThreadLocal` for object pooling is disabled,
- a web-safe implementation of
link:../javadoc/log4j-api/org/apache/logging/log4j/spi/ThreadContextMap.html[`ThreadContextMap`] is used,
- JMX notifications are sent synchronously,
- the JVM shutdown hook is disabled.

See xref:manual/systemproperties.adoc#log4j2.isWebapp[`log4j2.isWebapp`] for more details.
====

Using a logging implementation like **Log4j Core** in a Jakarta EE application requires particular care.
Since the lifecycle of a container or web application is independent of the lifecycle of the JVM, it's important for logging resources to be properly cleaned up (database connections closed, files closed, etc.) when the container or web application shuts down.

To properly synchronize the lifecycles of Log4j Core and Jakarta EE applications an additional **Log4j Web** artifact is provided.

[#log4j-jakarta-web-installation]

> === Installation

Reviewer: Can you add an ID to all sections, please?

Author: Done in 6ad209e057421baa57a79de2fe594a1f92d07b51.

> In order to install Log4j Web in your Web application, you need to add it as runtime dependency:

Reviewer: ```suggestion
In order to install Log4j Web in your web application, you need to add it as runtime dependency:
```

include::partial$manual/dependencies-log4j-jakarta-web.adoc[]

If you are writing a Servlet 3.0 or later application, Apache Log4j Web will register a
https://jakarta.ee/specifications/servlet/5.0/apidocs/jakarta/servlet/servletcontainerinitializer[`ServletContainerInitializer`]
that takes care of configuring the Log4j lifecycle for you. Under the hood this will:

* initialize Log4j Core with the correct configuration file,
* register a `Log4jServletContextListener` to automatically shut down Log4j Core, when the application shuts down,
* register a `Log4jServletFilter` to enable the xref:manual/lookups.adoc#WebLookup[web lookup].

See also <<web-application-specific>>.

[WARNING]
====
While the Servlet Specification allows web fragments to automatically add context listeners, it does not give any guarantees regarding the order in which those listeners are executed

> (cf. https://jakarta.ee/specifications/servlet/5.0/jakarta-servlet-spec-5.0#Assembling_the_descriptor[section 8.2.3]).

Reviewer: ```suggestion
(see https://jakarta.ee/specifications/servlet/5.0/jakarta-servlet-spec-5.0#Assembling_the_descriptor[Section 8.2.3]).
```

If other context listeners in your application use logging, you need to make sure that `Log4jServletContextListener` is the last listener to be executed at shutdown.

> In order to do it you must create a `web.xml` descriptor and add the `Log4jServletContextListener` explicitly as **first** context listener:

Reviewer: ```suggestion
In order to do it you must create a `web.xml` descriptor and add the `Log4jServletContextListener` explicitly as the **first** context listener:
```

[source,xml,indent=0]
----

> include::example$manual/webapp/web.xml[tag=context-listener]

Reviewer: Mind providing a link to the actual source in the caption of the source block, please?

Author: Fixed in 2809eaae3db7089f9c56b9b1e5a218e0606a4f38.

----
====

==== Manual installation

> If you are maintaining an older Servlet 2.5 (or earlier) application or if you disabled

Reviewer: Did you forget something here? :sweat_smile:

Author: Nice catch, fixed in 6ad209e057421baa57a79de2fe594a1f92d07b51.

> [source,xml,indent=0]

Reviewer: Mind providing a link to the actual source in the caption of the source block, please?

[id=isLog4jAutoInitializationDisabled]
==== `isLog4jAutoInitializationDisabled`

[cols="1h,5"]
|===
| Type          | `boolean`
| Default value | `false`
|===

If set to `true`, the `Log4jServletContainerInitializer` will be disabled, which prevents the automatic registration of both the `Log4jServletContextListener` and `Log4jServletFilter`.

[id=isLog4jAutoShutdownDisabled]
==== `isLog4jAutoShutdownDisabled`

[cols="1h,5"]
|===
| Type          | `boolean`
| Default value | `false`
|===

If set to `true`, the `Log4jServletContextListener` will not register a `Log4jServletContextListener` to handle the web application shut down.

[id=log4j.stop.timeout.timeunit]
==== `log4j.stop.timeout.timeunit`

[cols="1h,5"]
|===
| Type
| https://docs.oracle.com/javase/{java-target-version}/docs/api/java/util/concurrent/TimeUnit.html[`TimeUnit`]

| Default value
| https://docs.oracle.com/javase/{java-target-version}/docs/api/java/util/concurrent/TimeUnit.html#SECONDS[`SECONDS`]
|===

The `TimeUnit` to use for the shut-down delay.

[id=log4j.stop.timeout]
==== `log4j.stop.timeout`

[cols="1h,5"]
|===
| Type
| `long`

| Default value
| `30`
|===

> 

Reviewer: Mind adding some description, please?

====
The Log4j Taglib component is deprecated and is scheduled for removal in Log4j 3.

Currently, it only works with JavaServer Pages 2.3 or older and no version compatible with Jakarta Server Pages 3.0 is available.
====

The Log4j Taglib library defines a tag for most
link:../javadoc/log4j-api/org/apache/logging/log4j/Logger.html[`Logger`]
methods, including:

* simple and parameterized log statements:
+
[source,xml,indent=0]
----
include::example$manual/webapp/taglib.jsp[tag=simple]
----

* flow tracing statements:
+
[source,xml,indent=0]
----
include::example$manual/webapp/taglib.jsp[tag=entry-exit]
----

* catching and throwing statements:
+
[source,xml,indent=0]
----
include::example$manual/webapp/taglib.jsp[tag=catching]
----

* tags to test the current log level:
+
[source,xml,indent=0]
----
include::example$manual/webapp/taglib.jsp[tag=if-enabled]
----

* tags to set the name of the logger used:
+
[source,xml,indent=0]
----
include::example$manual/webapp/taglib.jsp[tag=set-logger]
----

* a `dump` tag that prints the contents of a JSP scope:
+
[source,xml,indent=0]
----
include::example$manual/webapp/taglib.jsp[tag=dump]
----

A complete JSP page example is available

> {antora-examples-url}/manual/webapp/taglib.jsp[on GitHub].

Reviewer: While it is personal, I very much like disclosing the entire source in the block caption, for instance,

```
. Snippet from an example {antora-examples-url}/manual/webapp/taglib.jsp[`taglib.jsp`]
[source,xml,indent=0]
----
include::example$manual/webapp/taglib.jsp[tag=dump]
----
```

Author: Fixed in 2809eaae3db7089f9c56b9b1e5a218e0606a4f38

+
Check also the <<web-application-wildfly,WildFly note above>>.

[WARNING]
====
Web application classloaders (see
https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0#web-application-class-loader[Servlet Specification 10.7.2]
) use a _"parent last"_ delegation strategy, but prevent application from overriding implementation classes provided by the container.

If you share Log4j between applications and the applications themselves contain Log4j Core, the logging behavior depends on the application server.
Some application servers will use the shared instance (e.g. WildFly), while others will use the application instance (e.g. Tomcat).
There are two solutions to this problem:

* you can remove Log4j from the WAR or EAR archive:

Maven::
If you use Maven, you can declare the
https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope[scope]
of all Log4j libraries as `provided`.

Gradle::
If you use Gradle, you can add `log4j-api` to the `providedCompile` configuration, while `log4j-core` to the `providedRuntime` configuration.

> See https://docs.gradle.org/current/userguide/war_plugin.html#sec:war_dependency_management[Gradle WAR plugin] for more details.

Reviewer: Mind placing these in tab block, please? :innocent: That will comply with the convention followed in the rest of the manual.

Author: Fixed in 2809eaae3db7089f9c56b9b1e5a218e0606a4f38.

system property to:

* either
link:../javadoc/log4j-core/org/apache/logging/log4j/core/selector/BasicContextSelector.html[`org.apache.logging.log4j.core.selector.BasicContextSelector`]
to use synchronous loggers,
* or
link:../javadoc/log4j-core/org/apache/logging/log4j/core/async/BasicAsyncLoggerContextSelector.html[`org.apache.logging.log4j.core.async.BasicAsyncLoggerContextSelector`]
to use asynchronous loggers.

In this approach you must use xref:manual/lookups.adoc[lookups] to register the application that generated a log event.
The most useful lookups in this case are:

Web lookup::
It does not require any setup, but it is available only after `Log4jServletFilter` has been executed.
Some log events pertinent to a web application can be unmarked.
See xref:manual/lookups.adoc#WebLookup[web lookup] for more information.

JNDI lookup::
It covers a larger part of the handling of a request, but it requires additional setup to export the name of the application via JNDI.
See xref:manual/lookups.adoc#JndiLookup[JNDI lookup] for more information.

When using a single logger context you choose between:

* logging all events to a single appender.

> You'll need a tool like Elastic Search to separate log events.
> We recommend the usage of a structured layout like
> xref:manual/json-template-layout.adoc[]
> in this case:

Reviewer: ```suggestion
We strongly recommend using a structured layout (e.g., xref:manual/json-template-layout.adoc[]) with an additional field capturing the Servlet context name.
This would allow separation of application logs by filtering on the context name.
Following example demonstrates this scheme using a Socket Appender writing to Elasticsearch:
```

The example Log4j configuration files need to be adapted too.

+
[tabs]
====
XML::
+
[source,xml,indent=0]
----
include::example$manual/webapp/log4j2-single.xml[tag=global]
----

JSON::
+
[source,json,indent=0]
----
include::example$manual/webapp/log4j2-single.json[tag=global]
----

YAML::
+
[source,yaml,indent=0]
----
include::example$manual/webapp/log4j2-single.yaml[tag=global]
----

Properties::
+
[source,properties,indent=0]
----
include::example$manual/webapp/log4j2-single.properties[tag=global]
----
====

* logging events to a separate appender for each application.
In this case you can use
xref:manual/appenders.adoc#RoutingAppender[routing appender]
to separate the events:
+
[tabs]
====
XML::
+
[source,xml,indent=0]
----

> include::example$manual/webapp/log4j2-single.xml[tag=routing]

Reviewer: I am still in favor of using JTL – that is what we recommend for production. (You can leave the `FileAppender` usage there.)

To overcome this limitation Log4j Core provides an alternative algorithm to determine the right logger context to choose: JNDI lookups.

[#jndi-configuration]
==== JNDI context selector

Application servers set up the correct JNDI context as soon as they determine, which application will handle a request.
Log4j Core allows the usage of JNDI to coordinate the usage of logger contexts in a Jakarta EE application server.
In order to use this feature, you need to:

. Set the
xref:manual/systemproperties.adoc#log4j2.contextSelector[`log4j2.contextSelector`]
Log4j configuration property to ``org.apache.logging.log4j.core.selector.JndiContextSelector``,
. For security reasons you need to enable the selector, by setting the
xref:manual/systemproperties.adoc#log4j2.enableJndiContextSelector[`log4j2.enableJndiContextSelector`]
Log4j configuration property to `true`,
. Each web application needs to configure the servlet context parameter `isLog4jContextSelectorNamed` to `true` and provide a value for the `log4jContextName` servlet context parameter and `java:comp/env/log4j/context-name` JNDI environment entry:
+
[source,xml,indent=0]
----

> include::example$manual/webapp/jndi.xml[tag=jndi]

Reviewer: Mind providing a link to the actual source in the caption of the source block, please?

Author: Fixed in 0338eada282faaa12a16cb27b9e0e6c2198e90e2.
